src/Build/Instance/HostServices.cs

```diff
@@ -84,23 +112,45 @@ public void RegisterHostObject(string projectFile, string targetName, string tas
             ErrorUtilities.VerifyThrowArgumentNull(targetName, "targetName");
             ErrorUtilities.VerifyThrowArgumentNull(taskName, "taskName");
 
-            // We can only set the host object to a non-null value if the affinity for the project is not out of proc, or if it is, it is only implicitly 
+            // We can only set the host object to a non-null value if the affinity for the project is not out of proc, or if it is, it is only implicitly
             // out of proc, in which case it will become in-proc after this call completes.  See GetNodeAffinity.
             bool isExplicit;
             bool hasExplicitOutOfProcAffinity = (GetNodeAffinity(projectFile, out isExplicit) == NodeAffinity.OutOfProc) && (isExplicit == true);
             ErrorUtilities.VerifyThrowInvalidOperation(!hasExplicitOutOfProcAffinity || hostObject == null, "InvalidHostObjectOnOutOfProcProject");
             _hostObjectMap = _hostObjectMap ?? new Dictionary<string, HostObjects>(StringComparer.OrdinalIgnoreCase);
 
-            HostObjects hostObjects;
-            if (!_hostObjectMap.TryGetValue(projectFile, out hostObjects))
-            {
-                hostObjects = new HostObjects();
-                _hostObjectMap[projectFile] = hostObjects;
-            }
+            HostObjects hostObjects = GetHostObjectsFromMapByKeyOrCreateNew(projectFile);
 
             hostObjects.RegisterHostObject(targetName, taskName, hostObject);
         }
 
+#if FEATURE_COM_INTEROP
+        /// <summary>
+        /// Register a remote host object for a particular task/target pair.
+        /// The remote host object require registered in Running Object Table(ROT) already.
+        /// Overwrites any existing host object.
+        ///
+        /// It's caller's responsibly:
+        /// To maintain the live cycle of the host object.
+        /// Register and unregister from ROT.
```

Comment:
Should we have a sanity check that the object is registered when we get to this point? Or is that too much?

Reply:
I did think about it. To verify registration, we should also make sure they have the correct interface that is used by the task(which we don't know). I think do half of a job is probably the same as no verification and give people false security. Plus we would need to add another COM API "IsRunning"

But I don't hold it strongly

Reply:
I think for now we don't add integration tests